<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine, text

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.
query = """
SELECT COUNT(*) 
FROM stores;
"""

def execute_SQL_query(query):
    with engine.connect() as conn:
      result = conn.execute(text(query))
      columns = result.keys() # Extract column names
      main_df = pd.DataFrame(result.fetchall(), columns=columns)
    print(main_df)

execute_SQL_query(query)

   count
0   1973


**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, name
FROM stores
ORDER BY name
"""

execute_SQL_query(sql)


      store                                name
0      4986  218 Fuel Express & Chubby's Liquor
1      3047          3047 Cub Foods / Iowa City
2      3060               3060 Cub Foods / Ames
3      3061         3061 Cub Foods / Sioux City
4      4845                  3rd St Convenience
...     ...                                 ...
1968   3992       Xo Food And Liquor / Waterloo
1969   4066                       Xpress Liquor
1970   4886                      Yr Dollar Stop
1971   4908                Zapf's Pronto Market
1972   4874                      Z's Quickbreak

[1973 rows x 2 columns]


**Prompt 2 Answer**: Store 3061, 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
sales.vendor,
sales.btl_price,
sales.bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 3
"""

execute_SQL_query(sql)

                      vendor btl_price  bottle_qty
0  Adamba Imports Int'l Inc.     $8.99          12
1      A Hardy / U.S.A. Ltd.    $28.50           3
2      A Hardy / U.S.A. Ltd.    $28.50          30


**Prompt 3 Answer**: (8.99, 12), (28.50, 6), ($28.50, 3))


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
SUM(btl_price * bottle_qty)
FROM sales
WHERE date >= '2015-02-01' AND date <= '2015-02-28'
"""

execute_SQL_query(sql)

              sum
0  $21,295,350.55


**Prompt 4 Answer**: $21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
VENDOR,
COUNT(*) AS total_txn,
SUM(bottle_qty) AS bottles_sold,
SUM(total) AS total_rev
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY total_txn DESC
"""

execute_SQL_query(sql)

                                  vendor  total_txn  bottles_sold   total_rev
0       Pernod Ricard USA/Austin Nichols      42338        403450  7880365.88
1                    Bacardi U.S.A. Inc.      26441        236047  5771242.07
2                        Diageo Americas      15725        152038  3346425.42
3        Constellation Wine Company Inc.      12565        116053  1895005.51
4                        Jim Beam Brands       5850        104119  1807097.23
5                            Stoli Group       4443         35513   832071.19
6                   Bacardi U.S.A., Inc.       3344         30239   765347.31
7                   Imperial Brands Inc.       2799         21033   265317.79
8                 Moet Hennessy USA Inc.       2153         10966   323126.54
9       Constellation Wine Company, Inc.       1672         12956   200283.64
10                               Proximo       1628          9431   178212.03
11                               MHW Ltd       1159          449

**Prompt 5 Answer**: 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
btl_price,
state_btl_cost,
btl_price - state_btl_cost AS Markup
FROM sales
LIMIT 100
"""

execute_SQL_query(sql)

   btl_price state_btl_cost  markup
0     $17.24         $11.49   $5.75
1     $17.24         $11.49   $5.75
2     $17.24         $11.49   $5.75
3     $25.73         $17.15   $8.58
4      $7.50          $5.00   $2.50
..       ...            ...     ...
95    $16.77         $11.18   $5.59
96    $17.25         $11.50   $5.75
97    $15.74         $10.49   $5.25
98    $86.99         $57.99  $29.00
99    $15.60         $10.40   $5.20

[100 rows x 3 columns]


### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
store,
AVG((btl_price - state_btl_cost)/state_btl_cost) as pct_markup
FROM sales
GROUP BY store
ORDER BY pct_markup DESC
"""

execute_SQL_query(sql)

      store  pct_markup
0      4013    0.525471
1      4024    0.521499
2      4266    0.518903
3      3833    0.515094
4      3660    0.514591
...     ...         ...
1347   9018    0.500000
1348   4784    0.499994
1349   4494    0.499909
1350   4534    0.499781
1351   4693    0.499214

[1352 rows x 2 columns]


**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
sales.store,
AVG((sales.btl_price - sales.state_btl_cost)/sales.state_btl_cost) as pct_markup,
stores.name,
MAX((sales.btl_price - sales.state_btl_cost)/sales.state_btl_cost) AS max_pct,
MIN((sales.btl_price - sales.state_btl_cost)/sales.state_btl_cost) AS min_pct
FROM sales
LEFT JOIN stores ON sales.store = stores.store
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY pct_markup DESC
"""

execute_SQL_query(sql)

      store  pct_markup                                  name   max_pct  \
0      4013    0.525471                Wal-Mart 0841 / Tipton  6.708763   
1      4024    0.521499            Wal-Mart 1546 / Iowa Falls  6.708763   
2      4266    0.518903            Wal-Mart 1683 / Shenandoah  6.708763   
3      3833    0.515094              Wal-Mart 3394 / Atlantic  6.708763   
4      3660    0.514591             Wal-Mart 2935 / Knoxville  6.708763   
...     ...         ...                                   ...       ...   
1347   9018    0.500000           Paradise Distilling Company  0.500000   
1348   4784    0.499994  Hawkeye Convenience Store Wiley Blvd  0.502488   
1349   4494    0.499909              SELECT MART / SIOUX CITY  0.572120   
1350   4534    0.499781                           D & B Foods  0.501235   
1351   4693    0.499214  Hawkeye Convenience Stores / 16th St  0.502415   

       min_pct  
0     0.498571  
1     0.498462  
2     0.498462  
3     0.498462  
4     0.498462

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * FROM
(
SELECT a.county,
(sum(b.bottle_qty)/a.population) AS bottles_per_capita,
ROUND(AVG(c.bottle_size), 1) AS avg_bottle_size
FROM counties a
LEFT JOIN sales b on a.county = b.county
LEFT JOIN products c on b.item = c.item_no
GROUP BY a.county
ORDER BY bottles_per_capita DESC
LIMIT 9
) e
LIMIT 1
"""

execute_SQL_query(sql)

      county  bottles_per_capita avg_bottle_size
0  Dickinson                  22           974.1


**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
btl_price,
bottle_qty,
btl_price * bottle_qty AS cal_total,
CAST(total AS MONEY) AS total
FROM sales
WHERE (btl_price * bottle_qty) <> CAST(total AS MONEY)
LIMIT 10
"""

execute_SQL_query(sql)

Empty DataFrame
Columns: [btl_price, bottle_qty, cal_total, total]
Index: []


**Prompt 10 Answer**: 0 transactions.